Added compilation steps

David-OConnor opened the pull request:
This adds explicit compilation steps to the `add-build-steps` documentation. Should make things easier on new users, and the hero who have been helping them on Discord and GH.

I'm unclear on how Mac fits in, so it might use a modification to clarify that.

Hwurzburg commented on dev/source/docs/building-the-code.rst:
```diff
@@ -15,18 +15,36 @@ Setting up the Build Environment
+Once the build environment is set up as above, run the following commmands prior to the initial build. (For Windows users, this is under WSL, as set up above):
 
-- Linux and MacOSX users should build with waf as described in `BUILD.md <https://github.com/ArduPilot/ardupilot/blob/master/BUILD.md>`__.
+- `git submodule init`
```
remove

David-OConnor replied:
It seems this is required

David-OConnor replied:
This is required to avoid the following error message:

```c
Build failed
 -> task in 'dronecan' failed (exit status 2):
        {task 140083940872752: dronecangen  -> }
 (run with -v to display more information)
 ```

Hwurzburg commented on dev/source/docs/building-the-code.rst:
```diff
@@ -15,18 +15,36 @@ Setting up the Build Environment
 
-- Linux and MacOSX users should build with waf as described in `BUILD.md <https://github.com/ArduPilot/ardupilot/blob/master/BUILD.md>`__.
+- `git submodule init`
+- `git submodule update --recursive`
```
```suggestion
- `git submodule update --recursive --init`
```

Hwurzburg commented on dev/source/docs/building-the-code.rst:
```diff
@@ -15,18 +15,36 @@ Setting up the Build Environment
-- Linux and MacOSX users should build with waf as described in `BUILD.md <https://github.com/ArduPilot/ardupilot/blob/master/BUILD.md>`__.
+- `git submodule init`
+- `git submodule update --recursive`
+- `./Tools/gittools/submodule-sync.sh`
```
this should only be used in case of emergency....its a sledge hammer....please remove it....we can add a note in the branching doc about using it in case of emergency where you cant get the modules updated correctly

David-OConnor replied:
It seems this is required for me.

khancyr replied:
not needed on fresh build

Hwurzburg commented on dev/source/docs/building-the-code.rst:
```diff
@@ -15,18 +15,36 @@ Setting up the Build Environment
+- `git submodule init`
+- `git submodule update --recursive`
+- `./Tools/gittools/submodule-sync.sh`
+- `python3 waf configure --board *BoardName*`
```
```suggestion
- `./waf configure --board *BoardName*`
```

David-OConnor replied:
Hi! It seems for me that `./waf` doesn't work. It is depending on how your system python is configured. Calling `python3` explicitly is more universal. (And works when I test it)

Hwurzburg commented on dev/source/docs/building-the-code.rst:
```diff
@@ -15,18 +15,36 @@ Setting up the Build Environment
+- `git submodule update --recursive`
+- `./Tools/gittools/submodule-sync.sh`
+- `python3 waf configure --board *BoardName*`
+- `python3 waf clean`
```
```suggestion
- `./waf clean`
```

David-OConnor replied:
See above: `./waf` may fail depending on system config.

Hwurzburg commented on dev/source/docs/building-the-code.rst:
```diff
@@ -15,18 +15,36 @@ Setting up the Build Environment
+`/ardupiot/libraries/AP_HAL_ChibiOS/hwdef`.
+
+Run the following command, each time you wish to build:
+`python3 waf copter`
```
```suggestion
`./waf copter`
```

David-OConnor commented:
Thank you for the review: After some experimenting based on your fixes, I committed most of your changes. These two items though, seem to be required:

```
- `git submodule update --recursive --init`
- `./Tools/gittools/submodule-sync.sh`
```

Ryanf55 commented:
> Thank you for the review: After some experimenting based on your fixes, I committed most of your changes. These two items though, seem to be required:
> 
> ```
> - `git submodule update --recursive --init`
> - `./Tools/gittools/submodule-sync.sh`
> ```

After running the `git submodule update --recursive --init`, can you share the output of 
`git submodule states` and `git status`?

Hwurzburg commented:
> > Thank you for the review: After some experimenting based on your fixes, I committed most of your changes. These two items though, seem to be required:
> > ```
> > - `git submodule update --recursive --init`
> > - `./Tools/gittools/submodule-sync.sh`
> > ```
> 
> After running the `git submodule update --recursive --init`, can you share the output of `git submodule status` and `git status`?

only the submodule-sync is unnecessary.....this is only if the module library has been re-orangized (not modified...update command takes care of that) since you updated the submodules

David-OConnor commented:
> > Thank you for the review: After some experimenting based on your fixes, I committed most of your changes. These two items though, seem to be required:
> > ```
> > - `git submodule update --recursive --init`
> > - `./Tools/gittools/submodule-sync.sh`
> > ```
> 
> After running the `git submodule update --recursive --init`, can you share the output of `git submodule status` and `git status`?

```c
david@Data-Desktop:~/ardupilot$ git submodule update --recursive --init
david@Data-Desktop:~/ardupilot$ git submodule status
 2696de2e47f8d83a3e4d24c4942cc4a8f5342cee modules/ChibiOS (heads/master)
 599965086437137ec0fe66e185611f43f335f889 modules/CrashDebug (heads/master)
 94ea1e98ffbb9d24b6717905e092dba2f211d1a1 modules/DroneCAN/DSDL (94ea1e9)
 ebaf96860a11a4cc43c01df6b651df143c6cde2d modules/DroneCAN/dronecan_dsdlc (heads/master)
 f9c584fe844af71a840d9b8d915b7a448c394ff6 modules/DroneCAN/libcanard (f9c584f)
 19fdf2e5b383243ccdb1094edae0603cf11469e8 modules/DroneCAN/pydronecan (1.0.16-2-g19fdf2e)
 3d1b17703c7cf4f22def2910bc845bdb5152d7b5 modules/Micro-CDR (heads/master)
 a2937bc5af606b2434e9a841adbf5c003fc1ec38 modules/Micro-XRCE-DDS-Client (heads/master)
 d572f4777349d43653b21d6c2fc63020ab326db2 modules/gbenchmark (v1.7.1)
 e1f690585d4803402584962bfaa8240ecaf1db30 modules/gsoap (heads/master)
 c5fed93f941865a0e912e9baf46ded713506590a modules/gtest (release-1.8.0-2268-gc5fed93f)
 9c39822c0f6bef4e7ecb5c4f95979f06c23f763d modules/mavlink (1.0.12-331-g9c39822c)
 1b1625b8e7da6e1307d73335cb995fa8813d5950 modules/waf (waf-1.9.2-731-g1b1625b8)
david@Data-Desktop:~/ardupilot$
```

```c
david@Data-Desktop:~/ardupilot$ git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   libraries/AP_HAL_ChibiOS/hwdef/MatekH743/hwdef-bl.dat
        modified:   libraries/AP_HAL_ChibiOS/hwdef/MatekH743/hwdef.dat

no changes added to commit (use "git add" and/or "git commit -a")
david@Data-Desktop:~/ardupilot$
```

khancyr commented on dev/source/docs/building-the-code.rst:
```diff
@@ -15,18 +15,42 @@ Setting up the Build Environment
+Once the build environment is set up as above, run the following commmands prior to the initial build. (For Windows users, this is using WSL, as set up above):
 
-- Linux and MacOSX users should build with waf as described in `BUILD.md <https://github.com/ArduPilot/ardupilot/blob/master/BUILD.md>`__.
+- `git clone https://github.com/ArduPilot/ardupilot`
```
we have setup the enviroement in previous setup, so no need to git clone again

khancyr commented on dev/source/docs/building-the-code.rst:
```diff
@@ -15,18 +15,42 @@ Setting up the Build Environment
-- Linux and MacOSX users should build with waf as described in `BUILD.md <https://github.com/ArduPilot/ardupilot/blob/master/BUILD.md>`__.
+- `git clone https://github.com/ArduPilot/ardupilot`
+- `cd ardupilot`
+- `git submodule update --recursive --init`
```
we shouldn't need this as previous instruction have done it already

TunaLobster replied:
A link to those instructions here would suffice.

hendjoshsr71 replied:
I disagree.

Somewhere on the welcome to dev with ardupilot we must show how to use the basics of waf.

That little link to the giant list in build.md is easy to miss and has too much detail...
We need a cohesive easy to follow story. 

I originally figured out how to use waf because I watched the commands  the ArduPilot extension was producing.

khancyr commented on dev/source/docs/building-the-code.rst:
```diff
@@ -15,18 +15,42 @@ Setting up the Build Environment
 
-.. youtube:: lNSvAPZOM_o
+`*BoardName*` above is the name of the board, as labeled by its associated folder in
+`/ardupiot/libraries/AP_HAL_ChibiOS/hwdef`.
```
no, board name isn't restricted to Chibios boards.
`./waf list_boards` show all the boards

Hwurzburg commented:
I gave this another look....carefully walking thru the existing wiki path for setting the build environment and then building the code, I am afraid this PR does not add anything but duplication.....I am willing to hear an argument against that...but am leaning on closing this as unnecessary duplicaton...

it does highlight the need to perhaps expand the "contributing code" section on the use of submodule updates and waf clean and the submodules-sync script....since those only are needed (assuming one follows the wiki in setup and building)  if you create or rebase local branches for code contribution and modules have gotten out of date, etc....but not just building from a properly cloned master, which the wiki leads you thru in the "Building the code" section

David-OConnor commented:
High. Going through as a new user, a list of commands needed to compile is important. Let's walk through the build steps on Windows using this page:

- Link to *Setup the Build Environment on Windows*:  This instructs the user to install WSL, or Cygwin.
- "Windows users: Windows users should fellow the directions described on Setup Build Environment on WIndows"

Do you see the problem? There are no Windows build instructions here. If you compare it to the page post this PR, a clear set of instructions are provided for all OSes. Ie, this PR makes it clear you need to build from Linux, and tells the user to explicitly follow the Linux instructions.

Addtionally, for the Linux page in `BUILD.md`, there are many steps. It's quite complicated. This PR adds a streamlined instructions, without navigating to a different page, of how to build. In my opinion, the page altered by the instructions clear for Windows, Linux and Mac users.

MichelleRos commented:
@David-OConnor 
> Do you see the problem? There are no Windows build instructions here!
I think there's two different things here - setting up the build environment, and compiling.

For the setup, this is actually correct... ArduPilot can't be built on Windows at all, which is why step one to build ArduPilot on Windows is essentially "install some form of Linux VM".

For compiling, the next step after the Setup the Build Environment on Windows page you mentioned actually explains how to setup VS Code to have "tasks" to build for various boards. I just took the alternative method, which is to just use the Ubuntu commands since WSL is essentially just an Ubuntu VM. I would just add a note to that page mentioning the VS Code setup (https://ardupilot.org/dev/docs/building-setup-windows.html) saying people could alternatively just follow the Ubuntu instructions if they prefer command-line.

David-OConnor commented:
The build instructions for windows are *build in an emulated Ubuntu*. To an experienced developer, this is clear. To a new person, who the docs are for, this should be spelled out. I think IDE choice is orthogonal; there is a recipe for building from Windows, and it's what's in this PR.

MichelleRos commented:
> * This links to a page that links to 2 sub-pages:
>   -- https://ardupilot.org/dev/docs/building-setup-windows10.html#building-setup-windows10: This provides instructions on how to install WSL, if not already installed.

It tells you how to set up WSL and install Ubuntu within it, yes, but then at step 3 on that page it says
> From the Start menu, start the “Ubuntu” application and then follow the [Ubuntu instructions to install ArduPilot development environment](https://ardupilot.org/dev/docs/building-setup-linux.html#building-setup-linux)
which is how you then setup & build ArduPilot within that Ubuntu.
